### PR TITLE
[Merged by Bors] - feat: Fintype.prod_prod_type

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Fin.lean [hunk 1]
```diff
@@ -484,6 +484,32 @@ theorem prod_ofFn {n : ℕ} {f : Fin n → α} : (ofFn f).prod = ∏ i, f i := b
 #align list.prod_of_fn List.prod_ofFn
 #align list.sum_of_fn List.sum_ofFn
 
+@[to_additive]
+theorem prod_finProd {n : ℕ} {m : ℕ} {f : ℕ × ℕ → α} :
+    ∏ x : Fin n × Fin m, f (x.1, x.2) = ∏ x : Fin n, ∏ y : Fin m, f (x, y) :=
+  prod_finset_product univ univ (fun _ ↦ univ) <|
+      fun p => ⟨fun _ => ⟨mem_univ p.1, mem_univ p.2⟩, fun _ => mem_univ p⟩
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Fin.lean:491]: This should be
```suggestion
@[to_additive Fintype.sum_prod]
theorem Fintype.prod_prod {α β} [Fintype α] [Fintype β] [CommMonoid γ] {f : α × β → γ} :
    ∏ x : α × β, f x = ∏ x : α, ∏ y : β, f (x, y) :=
  prod_finset_product univ univ (fun _ ↦ univ) <|
      fun p => ⟨fun _ => ⟨mem_univ p.1, mem_univ p.2⟩, fun _ => mem_univ p⟩
```
and belongs in a different file. I'm pretty sure we already have it soemwhere...

Comment (mo271) [reply Mathlib/Algebra/BigOperators/Fin.lean:491]: ok, great

Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Fin.lean:491]: Are you planning to open a new PR that adds this in the right place? Or did you find the lemma after all?

Comment (mo271) [reply Mathlib/Algebra/BigOperators/Fin.lean:491]: Oh, I just trusted you that we have it somewhere and was planning to look later where exactly. If I don't find it, I will re-open this pull request, moving and adapting the files...

Comment (mo271) [reply Mathlib/Algebra/BigOperators/Fin.lean]: I didn't find it (but that could also be lack of searching skills).

So I added it in the place you suggested. 
One thing still to fix is the fact that `Fintype (α₁ × α₂)` is not interred automatically...

File: Mathlib/Data/Fintype/BigOperators.lean [hunk 2]
```diff
@@ -295,4 +295,33 @@ theorem Fintype.prod_sum_type (f : Sum α₁ α₂ → M) :
 #align fintype.prod_sum_type Fintype.prod_sum_type
 #align fintype.sum_sum_type Fintype.sum_sum_type
 
+-- TODO: make sure this is inferred automatically from [Fintype α₁] and [Fintype α₂]
+variable [Fintype (α₁ × α₂)]
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/BigOperators.lean:299]: This must be defined in a different file; meaning you need to import something

File: Mathlib/Data/Fintype/BigOperators.lean [hunk 3]
```diff
@@ -295,4 +295,33 @@ theorem Fintype.prod_sum_type (f : Sum α₁ α₂ → M) :
...
+theorem Fintype.prod_prod_type' [CommMonoid γ] {f : α₁ → α₂ → γ} :
+    ∏ x : α₁ × α₂, f x.1 x.2 = ∏ x : α₁, ∏ y : α₂, f x y :=
+  prod_finset_product univ univ (fun _ ↦ univ) <|
+      fun p => ⟨fun _ => ⟨mem_univ p.1, mem_univ p.2⟩, fun _ => mem_univ p⟩
+
+@[to_additive Fintype.sum_prod_type_right]
+theorem Fintype.prod_prod_type_right [CommMonoid γ] {f : α₁ × α₂ → γ} :
+    ∏ x : α₁ × α₂, f x = ∏ y : α₂, ∏ x : α₁, f (x, y) :=
+  prod_finset_product_right univ univ (fun _ ↦ univ) <|
+      fun p => ⟨fun _ => ⟨mem_univ p.2, mem_univ p.1⟩, fun _ => mem_univ p⟩
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/BigOperators.lean:318]: There should be a shorter proof via `prod_comm`.

Comment (eric-wieser) [reply Mathlib/Data/Fintype/BigOperators.lean:318]: (which is perhaps an argument for skipping this lemma entirely)

File: Mathlib/Algebra/BigOperators/Fin.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Fin.lean]: Maybe @YaelDillies  can tell us where this lemma is hiding.

File: Mathlib/Data/Fintype/BigOperators.lean (hunk 2 above)
Comment (mo271) [reply Mathlib/Data/Fintype/BigOperators.lean]: done, imported `Data.Fintype.Prod`

File: Mathlib/Algebra/BigOperators/Fin.lean (hunk 1 above)
Comment (mo271) [reply Mathlib/Algebra/BigOperators/Fin.lean]: https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/group.20sum.2Fprod.20in.20pairs.20even.20and.20odd.20summands

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Fin.lean]: I mean this is defeq to `Finset.prod_product`.

Comment (mo271) [reply Mathlib/Algebra/BigOperators/Fin.lean]: Does this mean we want to have those nonetheless?

Comment (mo271) [reply Mathlib/Algebra/BigOperators/Fin.lean]: There is a slight inconsistency: `Finset.prod_product` has implicit arguments only, while `Finset.prod_disj_sum` has three explicit arguments. Don't know if this is intentional.

File: Mathlib/Data/Fintype/BigOperators.lean (hunk 3 above)
Comment (mo271) [reply Mathlib/Data/Fintype/BigOperators.lean]: I propose we go either with what I have now, or remove the `_right` versions. Right now `prod_product_right` has one real usage and `prod_product_right'` has none.

File: Mathlib/Data/Fintype/BigOperators.lean [hunk 4]
```diff
@@ -296,34 +296,26 @@ theorem Fintype.prod_sum_type (f : Sum α₁ α₂ → M) :
...
     ∏ x : α₁ × α₂, f x.1 x.2 = ∏ x : α₁, ∏ y : α₂, f x y :=
-  prod_prod_type
+  Finset.prod_product'
 
 @[to_additive Fintype.sum_prod_type_right]
 theorem Fintype.prod_prod_type_right [CommMonoid γ] {f : α₁ × α₂ → γ} :
-    ∏ x : α₁ × α₂, f x = ∏ y : α₂, ∏ x : α₁, f (x, y) := by
-  rw [prod_comm]
-  exact prod_prod_type
+    ∏ x : α₁ × α₂, f x = ∏ y : α₂, ∏ x : α₁, f (x, y) :=
```
Comment (YaelDillies) [inline Mathlib/Data/Fintype/BigOperators.lean:312]:
```suggestion
    ∏ x, f x = ∏ y, ∏ x, f (x, y) :=
```
and elsewhere

Comment (mo271) [reply Mathlib/Data/Fintype/BigOperators.lean]: done, also elsewhere, expect in some places in the uncurried versions, where it didn't work.

File: Mathlib/Data/Fintype/BigOperators.lean [hunk 5]
```diff
@@ -295,4 +296,26 @@ theorem Fintype.prod_sum_type (f : Sum α₁ α₂ → M) :
 #align fintype.prod_sum_type Fintype.prod_sum_type
 #align fintype.sum_sum_type Fintype.sum_sum_type
 
+@[to_additive (attr := simp) Fintype.sum_prod_type ]
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/BigOperators.lean:299]:
```suggestion
@[to_additive (attr := simp) Fintype.sum_prod_type]
```

**Comment by eric-wieser** [comment]
bors merge

Thanks!

**Pull request closed without merging**